File: dqops/src/main/java/com/dqops/execution/statistics/jobs/CollectStatisticsQueueJobParameters.java
/*
 * Copyright © 2021 DQOps (support@dqops.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dqops.execution.statistics.jobs;

import com.dqops.data.statistics.factory.StatisticsDataScope;
import com.dqops.execution.statistics.progress.SilentStatisticsCollectorExecutionProgressListener;
import com.dqops.execution.statistics.progress.StatisticsCollectorExecutionProgressListener;
import com.dqops.metadata.search.StatisticsCollectorSearchFilters;
import com.dqops.utils.exceptions.DqoRuntimeException;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.EqualsAndHashCode;

/**
 * Parameters object for the run the statistics collection job.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@EqualsAndHashCode(callSuper = false)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
public class CollectStatisticsQueueJobParameters implements Cloneable {
    /**
     * Statistics collectors search filters that identify the type of statistics collector to run.
     */
    @JsonPropertyDescription("Statistics collectors search filters that identify the type of statistics collector to run.")
    private StatisticsCollectorSearchFilters statisticsCollectorSearchFilters;

    /**
     * The target scope of collecting statistics. Statistics can be collected for the entire table or for each data stream separately.
     */
    @JsonPropertyDescription("The target scope of collecting statistics. Statistics can be collected for the entire table or for each data grouping separately.")
    private StatisticsDataScope dataScope = StatisticsDataScope.table;

    /**
     * Progress listener that will receive events during the statistics collection.
     */
    @JsonIgnore
    @EqualsAndHashCode.Exclude
    private StatisticsCollectorExecutionProgressListener progressListener = new SilentStatisticsCollectorExecutionProgressListener();

    /**
     * Boolean flag that enables a dummy statistics collection (sensors are executed, but the statistics results are not written to the parquet files).
     */
    @JsonPropertyDescription("Boolean flag that enables a dummy statistics collection (sensors are executed, but the statistics results are not written to the parquet files).")
    private boolean dummySensorExecution;

    /**
     * The summary of the statistics collection job after if finished. Returns the number of collectors analyzed, columns analyzed, statistics results captured.
     */
    @JsonPropertyDescription("The summary of the statistics collection job after if finished. Returns the number of collectors analyzed, columns analyzed, statistics results captured.")
    private CollectStatisticsResult collectStatisticsResult;

    public CollectStatisticsQueueJobParameters() {
    }

    /**
     * Creates statistics collection run parameters.
     * @param statisticsCollectorSearchFilters Statistics collector search filters.
     * @param progressListener Progress listener to receive events during the statistics collector execution.
     * @param dataScope Statistics data scope to analyze - the whole table or each data stream separately.
     * @param dummySensorExecution True when it is a dummy run, only for showing rendered sensor queries.
     */
    public CollectStatisticsQueueJobParameters(StatisticsCollectorSearchFilters statisticsCollectorSearchFilters,
                                               StatisticsCollectorExecutionProgressListener progressListener,
                                               StatisticsDataScope dataScope,
                                               boolean dummySensorExecution) {
        this.statisticsCollectorSearchFilters = statisticsCollectorSearchFilters;
        this.progressListener = progressListener;
        this.dataScope = dataScope;
        this.dummySensorExecution = dummySensorExecution;
    }

    /**
     * Returns the statistics collectors search filters.
     * @return Statistics collectors search filters.
     */
    public StatisticsCollectorSearchFilters getStatisticsCollectorSearchFilters() {
        return statisticsCollectorSearchFilters;
    }

    /**
     * Progress listener that should be used to run the statistics collectors.
     * @return Statistics collectors progress listener.
     */
    public StatisticsCollectorExecutionProgressListener getProgressListener() {
        return progressListener;
    }

    /**
     * Returns the statistics data scope (whole table or each data stream separately).
     * @return Data scope configuration.
     */
    public StatisticsDataScope getDataScope() {
        return dataScope;
    }

    /**
     * Returns true if it should be just a dummy run without actually running any queries.
     * @return true when it is a dummy run.
     */
    public boolean isDummySensorExecution() {
        return dummySensorExecution;
    }

    /**
     * Returns the results of a finished job with the count of statistics that were collected.
     * @return The results of a finished job with the count of statistics that were collected.
     */
    public CollectStatisticsResult getCollectStatisticsResult() {
        return collectStatisticsResult;
    }

    /**
     * Sets the summary result of the statistics collection job, this method is called by the job when it finished successfully.
     * @param collectStatisticsResult The statistics collection result.
     */
    public void setCollectStatisticsResult(CollectStatisticsResult collectStatisticsResult) {
        this.collectStatisticsResult = collectStatisticsResult;
    }

    /**
     * Creates and returns a copy of this object.
     */
    @Override
    public CollectStatisticsQueueJobParameters clone() {
        try {
            return (CollectStatisticsQueueJobParameters)super.clone();
        }
        catch (CloneNotSupportedException ex) {
            throw new DqoRuntimeException("Clone not supported", ex);
        }
    }
}

File: distribution/python/dqops/client/models/collect_statistics_queue_job_parameters.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..models.statistics_data_scope import StatisticsDataScope
from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.collect_statistics_result import CollectStatisticsResult
    from ..models.statistics_collector_search_filters import (
        StatisticsCollectorSearchFilters,
    )


T = TypeVar("T", bound="CollectStatisticsQueueJobParameters")


@_attrs_define
class CollectStatisticsQueueJobParameters:
    """
    Attributes:
        statistics_collector_search_filters (Union[Unset, StatisticsCollectorSearchFilters]):
        data_scope (Union[Unset, StatisticsDataScope]):
        dummy_sensor_execution (Union[Unset, bool]): Boolean flag that enables a dummy statistics collection (sensors
            are executed, but the statistics results are not written to the parquet files).
        collect_statistics_result (Union[Unset, CollectStatisticsResult]): Returns the result with the summary of the
            statistics collected.
    """

    statistics_collector_search_filters: Union[
        Unset, "StatisticsCollectorSearchFilters"
    ] = UNSET
    data_scope: Union[Unset, StatisticsDataScope] = UNSET
    dummy_sensor_execution: Union[Unset, bool] = UNSET
    collect_statistics_result: Union[Unset, "CollectStatisticsResult"] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        statistics_collector_search_filters: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.statistics_collector_search_filters, Unset):
            statistics_collector_search_filters = (
                self.statistics_collector_search_filters.to_dict()
            )

        data_scope: Union[Unset, str] = UNSET
        if not isinstance(self.data_scope, Unset):
            data_scope = self.data_scope.value

        dummy_sensor_execution = self.dummy_sensor_execution
        collect_statistics_result: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.collect_statistics_result, Unset):
            collect_statistics_result = self.collect_statistics_result.to_dict()

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if statistics_collector_search_filters is not UNSET:
            field_dict["statistics_collector_search_filters"] = (
                statistics_collector_search_filters
            )
        if data_scope is not UNSET:
            field_dict["data_scope"] = data_scope
        if dummy_sensor_execution is not UNSET:
            field_dict["dummy_sensor_execution"] = dummy_sensor_execution
        if collect_statistics_result is not UNSET:
            field_dict["collect_statistics_result"] = collect_statistics_result

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.collect_statistics_result import CollectStatisticsResult
        from ..models.statistics_collector_search_filters import (
            StatisticsCollectorSearchFilters,
        )

        d = src_dict.copy()
        _statistics_collector_search_filters = d.pop(
            "statistics_collector_search_filters", UNSET
        )
        statistics_collector_search_filters: Union[
            Unset, StatisticsCollectorSearchFilters
        ]
        if isinstance(_statistics_collector_search_filters, Unset):
            statistics_collector_search_filters = UNSET
        else:
            statistics_collector_search_filters = (
                StatisticsCollectorSearchFilters.from_dict(
                    _statistics_collector_search_filters
                )
            )

        _data_scope = d.pop("data_scope", UNSET)
        data_scope: Union[Unset, StatisticsDataScope]
        if isinstance(_data_scope, Unset):
            data_scope = UNSET
        else:
            data_scope = StatisticsDataScope(_data_scope)

        dummy_sensor_execution = d.pop("dummy_sensor_execution", UNSET)

        _collect_statistics_result = d.pop("collect_statistics_result", UNSET)
        collect_statistics_result: Union[Unset, CollectStatisticsResult]
        if isinstance(_collect_statistics_result, Unset):
            collect_statistics_result = UNSET
        else:
            collect_statistics_result = CollectStatisticsResult.from_dict(
                _collect_statistics_result
            )

        collect_statistics_queue_job_parameters = cls(
            statistics_collector_search_filters=statistics_collector_search_filters,
            data_scope=data_scope,
            dummy_sensor_execution=dummy_sensor_execution,
            collect_statistics_result=collect_statistics_result,
        )

        collect_statistics_queue_job_parameters.additional_properties = d
        return collect_statistics_queue_job_parameters

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
